enclave/core/sgx/calls.c

```diff
@@ -279,15 +280,41 @@ oe_result_t oe_handle_call_enclave_function(uint64_t arg_in)
         args.output_buffer_size,
         &output_bytes_written);
 
-    // The output_buffer is expected to point to a marshaling struct,
-    // whose first field is an oe_result_t. The function is expected
-    // to fill this field with the status of the ecall.
-    result = *(oe_result_t*)output_buffer;
+    /*
+     * The output_buffer is expected to point to a marshaling struct.
+     * The function is expected to fill the struct.
+     */
+    return_args_ptr = (oe_call_function_return_args_t*)output_buffer;
 
+    result = return_args_ptr->result;
     if (result == OE_OK)
     {
+        /*
+         * Nonzero deepcopy_out_buffer and deepcopy_out_buffer_size fieldes
+         * indicate that there is deep-copied content that needs to be
+         * transmitted to the host. Also, ensuring that the content lies on
+         * enclave memory.
+         */
+        if (return_args_ptr->deepcopy_out_buffer &&
+            return_args_ptr->deepcopy_out_buffer_size &&
```

Comment:
In case there is a buffer, it would be better to error out if size == 0 or if the buffer does not lie within the enclave.